bpo-32030: Split Py_Main() into subfunctions

vstinner opened the pull request:
* Don't use "Python runtime" anymore to parse command line options or
  to get environment variables: pymain_init() is now a strict
  separation.
* Use an error message rather than "crashing" directly with
  Py_FatalError(). Limit the number of calls to Py_FatalError(). It
  prepares the code to handle errors more nicely later.
* Warnings options (-W, PYTHONWARNINGS) and "XOptions" (-X) are now
  only added to the sys module once Python core is properly
  initialized.
* _PyMain is now the well identified owner of some important strings
  like: warnings options, XOptions, and the "program name". The
  program name string is now properly freed at exit.
  pymain_free() is now responsible to free the "command" string.
* Rename most methods in Modules/main.c to use a "pymain_" prefix to
  avoid conflits and ease debug.
* Replace _Py_CommandLineDetails_INIT with memset(0)
* Reorder a lot of code to fix the initialization ordering. For
  example, initializing standard streams now comes before parsing
  PYTHONWARNINGS.
* pymain_init() now uses the cmdline structure rather than relying on
  Py_xxx global configuration variables. For example, replace
  Py_UnbufferedStdioFlag with cmdline.use_unbuffered_io.
* Py_Main() now handles errors when adding warnings options and
  XOptions.
* Add _PyMem_GetDefaultRawAllocator() private function.

<!-- issue-number: bpo-32030 -->
https://bugs.python.org/issue32030
<!-- /issue-number -->

vstinner committed:
bpo-32030: Split Py_Main() into subfunctions

* Don't use "Python runtime" anymore to parse command line options or
  to get environment variables: pymain_init() is now a strict
  separation.
* Use an error message rather than "crashing" directly with
  Py_FatalError(). Limit the number of calls to Py_FatalError(). It
  prepares the code to handle errors more nicely later.
* Warnings options (-W, PYTHONWARNINGS) and "XOptions" (-X) are now
  only added to the sys module once Python core is properly
  initialized.
* _PyMain is now the well identified owner of some important strings
  like: warnings options, XOptions, and the "program name". The
  program name string is now properly freed at exit.
  pymain_free() is now responsible to free the "command" string.
* Rename most methods in Modules/main.c to use a "pymain_" prefix to
  avoid conflits and ease debug.
* Replace _Py_CommandLineDetails_INIT with memset(0)
* Reorder a lot of code to fix the initialization ordering. For
  example, initializing standard streams now comes before parsing
  PYTHONWARNINGS.
* Py_Main() now handles errors when adding warnings options and
  XOptions.
* Add _PyMem_GetDefaultRawAllocator() private function.
* Cleanup _PyMem_Initialize(): remove useless global constants: move
  them into _PyMem_Initialize().
* Call _PyRuntime_Initialize() as soon as possible:
  _PyRuntime_Initialize() now returns an error message on failure.
* Add _PyInitError structure and following macros:

  * _Py_INIT_OK()
  * _Py_INIT_ERR(msg)
  * _Py_INIT_USER_ERR(msg): "user" error, don't abort() in that case
  * _Py_INIT_FAILED(err)

vstinner commented:
Hum, test_capi fails on Windows.

vstinner commented:
test_capi failed on the "gcc C" job of Travis CI:

```
0:01:20 load avg: 44.72 [ 46/404] test_capi
*** Error in `/home/travis/build/python/cpython/venv/bin/python': munmap_chunk(): invalid pointer: 0x00007f0f5c5e90a0 ***
*** Error in `/home/travis/build/python/cpython/venv/bin/python': munmap_chunk(): invalid pointer: 0x00007f26c992e0a0 ***
*** Error in `/home/travis/build/python/cpython/venv/bin/python': munmap_chunk(): invalid pointer: 0x00007fab599500a0 ***
test test_capi failed -- multiple errors occurred; run in verbose mode for details
```

ncoghlan commented on Include/pylifecycle.h:
```diff
@@ -21,17 +21,17 @@ PyAPI_FUNC(int) Py_SetStandardStreamEncoding(const char *encoding,
 
 /* PEP 432 Multi-phase initialization API (Private while provisional!) */
-PyAPI_FUNC(void) _Py_InitializeCore(const _PyCoreConfig *);
+PyAPI_FUNC(const char*) _Py_InitializeCore(const _PyCoreConfig *);
```
I'm wondering if a `typedef char _Py_InitErrorMessage;` might make these API signatures a bit clearer:

    PyAPI_FUNC(const _Py_InitErrorMessage*) _Py_InitializeCore(const _PyCoreConfig *);

While I currently guessed "Oh, that will be an error message" in the context of the patch, I'm not sure I would have without that extra context.

Using a typedef also leaves the door open to upgrading to a struct later if we find a reason for doing that.

vstinner replied:
I didn't see the need for a structure, but after converting more code to avoid calling Py_FatalError(), I found two places where we exit Python but don't want to exit with abort():

```
        if (Py_ReadHashSeed(seed_text, &use_hash_seed, &hash_seed) < 0) {
            Py_FatalError("PYTHONHASHSEED must be \"random\" or an integer "
                          "in range [0; 4294967295]");
        }
```

and

```
    m = PyImport_ImportModule("site");
    if (m == NULL) {
        fprintf(stderr, "Failed to import the site module\n");
        PyErr_Print();
        Py_Finalize();
        exit(1);
    }
```

These errors are more on the user side, they are not "bugs" of Python itself.

vstinner replied:
So I implemented a new _PyInitError "type" (technically a structure) to indicate if it's an "user error" or another error. Moreover, I added the name of the function where the error was raised, to the error message.

ncoghlan commented on Modules/main.c:
```diff
@@ -35,9 +35,11 @@
 extern "C" {
 #endif
 
+extern void _PyMem_GetDefaultRawAllocator(PyMemAllocatorEx *alloc);
```
When I used this technique for the locale coercion APIs, I ended up causing a build failure on Cygwin: https://bugs.python.org/issue31877

Assuming that's a general concern, `PyAPI_FUNC` will be needed here as well.

vstinner replied:
Fixed

ncoghlan reviewed (commented):
Nice! This looks really good to me - just a couple of minor suggestions inline.

ncoghlan commented:
One general comment: while I didn't spot any particular instances of this going wrong, it's important to keep in mind that the various global flags (Py_IsolatedFlag, etc) are part of the embedding configuration API. This means that even as we add the more explicit configuration structs and use those to pass settings from Py_Main into the CPython runtime, we still need to account for other applications that are passing in the settings via the global variables instead (and either keep that model working, or else provide explicit porting guidance).

vstinner commented:
The PR compiles correctly on macOS and the following tests passed:

```
macbook:master haypo$ PYTHONMALLOC=pymalloc ./python.exe -m test -j0 test_capi test_cmd_line test_cmd_line_script
Run tests in parallel using 6 child processes
0:00:08 load avg: 2.30 [1/3] test_cmd_line passed
0:00:11 load avg: 2.35 [2/3] test_cmd_line_script passed
0:00:17 load avg: 2.49 [3/3] test_capi passed
All 3 tests OK.

Total duration: 17 sec
Tests result: SUCCESS
```

vstinner commented:
> One general comment: while I didn't spot any particular instances of this going wrong, it's important to keep in mind that the various global flags (Py_IsolatedFlag, etc) are part of the embedding configuration API. (...)

Ok, I changed my code to set Py_xxx flags as soon as possible, and use Py_xxx flags rather than using pymain.

vstinner commented:
> Hum, test_capi fails on Windows.

and

> test_capi failed on the "gcc C" job of Travis CI: (...

Oh, I expect PyMem_RawMalloc() to be usable directly, but this changed recently: _PyRuntime_Initialize() must now be called before the first call to PyMem_RawMalloc(). I fixed that as well.

vstinner committed:
Fix _Py_InitializeCore()

Call _PyRuntime_Initialize() and copy core config before calling
_PyMem_SetupAllocators().

vstinner committed:
python.vcxproj: define Py_BUILD_CORE

Programs/python.c now uses internals/pystate.h which requires
Py_BUILD_CORE.

zooba commented:
Haven't looked closely, but this looks like overall goodness!

If you want one extra case to keep in mind, see how I implemented my [spython.c](https://github.com/zooba/cpython/blob/pep551/Programs/spython.c#L479-L508) for my PEP 551 implementation. There's obviously no compatibility requirement, but I'd want to be able to keep doing the same things after this change.

vstinner commented:
@zooba: "Haven't looked closely, but this looks like overall goodness!"

Oh thanks. With @ncoghlan and you who know the topic, I'm now confident enough to merge my change (I already merged it ;-)).

"If you want one extra case to keep in mind, see how I implemented my spython.c  for my PEP 551 implementation. There's obviously no compatibility  requirement, but I'd want to be able to keep doing the same things after  this change."

My change shouldn't break this code. @ncoghlan: would you mind to double-check?

In the first version of this change, I overrided Py_xxx flags from the command lines, ignoring their previous value. @ncoghlan warned me about Py_xxx flags set when Python is embedded. So I fixed my code.
